ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Reader.hs

```diff
@@ -0,0 +1,500 @@
+
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadThrow
+
+import           Control.Tracer (contramap, traceWith)
+
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Ouroboros.Network.Block (ChainUpdate (..), HasHeader,
+                     HeaderHash, Point, SlotNo, blockPoint, castPoint,
+                     pointSlot)
+import           Ouroboros.Network.Chain (genesisPoint)
+
+import           Ouroboros.Consensus.Block (GetHeader (..), headerPoint)
+import           Ouroboros.Consensus.Util.STM (blockUntilJust)
+
+import           Ouroboros.Storage.ChainDB.API (ChainDbError (..), Reader (..),
+                     ReaderId)
+
+import qualified Ouroboros.Storage.ChainDB.Impl.ImmDB as ImmDB
+import qualified Ouroboros.Storage.ChainDB.Impl.Query as Query
+import           Ouroboros.Storage.ChainDB.Impl.Types
+
+
+-- | Check if the ChainDB is open. If not, throw a 'ClosedDBError'. Next,
+-- check whether the reader with the given 'ReaderId' still exists. If not,
+-- throw a 'ClosedReaderError'.
+--
+-- Otherwise, execute the given function on the 'ChainDbEnv' and 'ReaderState'
+-- 'TVar'.
+getReader
+  :: forall m blk r.
+     ( MonadSTM m
+     , MonadThrow (STM m)
+     , HasHeader blk
+     )
+  => ChainDbHandle m blk
+  -> ReaderId
+  -> (ChainDbEnv m blk -> TVar m (ReaderState m blk) -> m r)
+  -> m r
+getReader (CDBHandle varState) readerId f = do
+    (env, varRdr) <- atomically $ readTVar varState >>= \case
+      ChainDbClosed _env -> throwM $ ClosedDBError @blk
+      -- See the docstring of 'ChainDbReopening'
+      ChainDbReopening   -> error "ChainDB used while reopening"
+      ChainDbOpen    env ->
+        Map.lookup readerId <$> readTVar (cdbReaders env) >>= \case
+          Nothing  -> throwM $ ClosedReaderError @blk readerId
+          Just varRdr -> return (env, varRdr)
+    f env varRdr
+
+-- | Variant 'of 'getReader' for functions taking one argument.
+getReader1
+  :: forall m blk a r.
+     ( MonadSTM m
+     , MonadThrow (STM m)
+     , HasHeader blk
+     )
+  => ChainDbHandle m blk
+  -> ReaderId
+  -> (ChainDbEnv m blk -> TVar m (ReaderState m blk) -> a -> m r)
+  -> a -> m r
+getReader1 h rdrId f a =
+    getReader h rdrId (\env varReader -> f env varReader a)
+
+newReader
+  :: forall m blk.
+     ( MonadSTM   m
+     , MonadCatch m
+     , MonadThrow (STM m)
+     , HasHeader blk
+     , HasHeader (Header blk)
+     )
+  => ChainDbEnv    m blk
+  -> ChainDbHandle m blk
+     -- ^ We need the handle to pass it on to the 'Reader', so that each time
+     -- a function is called on the 'Reader', we check whether the ChainDB is
+     -- actually open.
+  -> m (Reader m blk (BlockOrHeader blk))
+newReader cdb@CDB{..} h = do
+    tipPoint <- atomically $ Query.getTipPoint cdb
+    readerState <- if tipPoint == genesisPoint
+      then return $ ReaderInMem rollState
```

Comment:
Mention that the reason for having this case is to avoid streaming `FromExclusive genesisPoint` `ToInclusive genesisPoint`, which doesn't work. This problem will (might?) go away when we change the bounds types to exclude this impossible cases.

Reply:
That was not the reason. The reason is that when the ChainDB is empty (and thus the ImmutableDB too), there is no point in opening an ImmutableDB iterator.